### PR TITLE
bpo-37224: Improve test__xxsubinterpreters.DestroyTests

### PR DESCRIPTION
**Pull request by aeros**
Adds an additional assertion check based on a race condition for `test__xxsubinterpreters.DestroyTests.test_still_running` discovered in the bpo issue.

https://bugs.python.org/issue37224

Automerge-Triggered-By: @ericsnowcurrently

### CODE REVIEW

File: Lib/test/test__xxsubinterpreters.py
Comment (ericsnowcurrently) [inline Lib/test/test__xxsubinterpreters.py]: I wouldn't worry about wrapping the line like this.  It's one of those cases where we break PEP 8 for practicality. :smile:

Also, consider giving this line another indent, so it is not in-line with the with statement's body.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

I left small comment about formatting, but that's it. :)

File: Lib/test/test__xxsubinterpreters.py
Comment (aeros) [reply Lib/test/test__xxsubinterpreters.py]: Yeah, I think it's much cleaner without the line wrapping in this case. But, since different core devs have different opinions on how strictly we should adhere to the PEP8 line width limit in the standard library, I just got in the habit of wrapping it by default once it exceeds 79 chars.

In your opinion, what do you think the "hard limit" should be as far as line width goes? In my own code, I typically start to think about wrapping at around 80-100 characters, and have 120 as a hard limit.

Fixed.

Comment (ericsnowcurrently) [reply Lib/test/test__xxsubinterpreters.py]: Just consider that a message intended for readers shouldn't itself have too long a line length, regardless of how long it is encoded in the code.  So at a certain length (e.g. 80, 72) either the user-facing message should be shortened or wrapped (with explicit `\n` characters).  If wrapped, then it would make sense to also wrap the string in the code.  Hopefully that makes sense.  Regardless, rarely should there be a need to have such a long user-facing message.

Comment (aeros) [reply Lib/test/test__xxsubinterpreters.py]: In this case, we're at ~65 characters for the actual user message (depending on how many chars the interpreter number takes up), so I think this one is okay. Thanks for the info!

File: Lib/test/test__xxsubinterpreters.py
Comment (ericsnowcurrently) [inline Lib/test/test__xxsubinterpreters.py]: Typically we want to line up additional "arg lines" with the first "arg line".  So in this case the `msg=` line should line up the first arg, right after the `assertTrue(`, like this:

```suggestion
                            msg=f"Interp {interp} should be running before destruction.")
```

File: Lib/test/test__xxsubinterpreters.py
Comment (ericsnowcurrently) [inline Lib/test/test__xxsubinterpreters.py]: Likewise with lining up "arg lines":

```suggestion
                                   msg=f"Should not be able to destroy interp {interp} while it's still running."):
```

Also, personally I prefer not to split the opening clause of a compound statement over multiple lines.  However, in this case it isn't so big a deal, though I'd still try to keep the colon as visually distinct as possible, like this:

```suggestion
                                   msg=f"Should not be able to destroy interp {interp} while it's still running."
                                   ):
```

However, I don't think that buys much.  I'll leave it to your discretion. :smile:

**Review by ericsnowcurrently — APPROVED** [review]
still LGTM :smile:

I did leave a couple more formatting-related comments.  Sorry for dragging this out.  My intention is to be instructive rather than proscriptive.  (If you don't find this helpful then please tell me.)  So don't feel like making any of these changes is a prerequisite for merging.  I'll give it a day before merging, to give you a chance to make any formatting changes you deem worth doing. :smile:

**Comment by ericsnowcurrently** [comment]
Also, I expect this needs a backport to the 3.8 branch.  Please confirm.

**Comment by ericsnowcurrently** [comment]
Also, thanks again for working on this! :smile:

**Comment by aeros** [comment]
> Also, I expect this needs a backport to the 3.8 branch. Please confirm.

Yep, I added the `needs backport to 3.8` label. If there are any merge conflicts (and @miss-islington is unable to resolve them), I'll sort it out manually with cherry pick.

**Comment by aeros** [comment]
> I did leave a couple more formatting-related comments. Sorry for dragging this out. My intention is to be instructive rather than proscriptive. (If you don't find this helpful then please tell me.)

They were definitely quite helpful! Although I have some experience contributing to standard library code changes, it's rather uncommon to encounter situations like this where multi-line arguments are necessary. But, it's great to know for future reference how to format it in a way that's PEP8 compliant and highly readable.

> Also, thanks again for working on this!

No problem, thanks for the review! (:

After this is merged, I'll very likely do more investigation into the root cause of the running failure. It can a rather time consuming process to dive into the internals and experiment with different potential solutions. But I've recently found a strong interest in the different implementations of concurrency in the standard library, including asyncio, concurrent.futures, multiprocessing, and most certainly with your upcoming subinterpreters module.

Not to go off on too far off-topic of the PR, but I think CPython has a lot of room for improvement and increased relevance in the world of concurrent programming and building scalable, distributed systems. Many people still unfortunately view Python as "best suited for simplistic, single-threaded applications", despite it's success with concurrent network IO in asyncio. As a result, I'd like to invest as much time as I reasonably can in improving our core models of concurrency, so that CPython can be more competitive with (or even supersede) other ecosystems in that domain.

Thanks for all of your hard work with the subinterpreters module! I'm very excited about the possibilities it will pave the way for with it's drastically increased isolation and potential performance implications for CPU-bound operations.

File: Lib/test/test__xxsubinterpreters.py
Comment (aeros) [reply Lib/test/test__xxsubinterpreters.py]: > Also, personally I prefer not to split the opening clause of a compound statement over multiple lines. However, in this case it isn't so big a deal, though I'd still try to keep the colon as visually distinct as possible, like this:

I added the first two suggestions, but left out the `):` suggestion mainly because there's been some recent controversy about whether it's PEP8 compliant: https://discuss.python.org/t/pep-8-clarify-if-multiline-argument-list-with-a-closing-on-a-separate-line-is-acceptable/2948/10.

**Pull request merged**
